teps/0048-pipeline-as-code.md

```diff
@@ -0,0 +1,214 @@
+---
+status: proposed
+title: Pipelines as Code
+creation-date: '2020-01-20'
+last-updated: '2020-01-20'
+authors: [chmouel, afrittoli]
+---
+
+# TEP-0048: Pipeline as Code
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+
+## Summary
+
+The pipelines as code technique emphasizes that the configuration of delivering
+pipelines that build, test and deploy our applications or infrastructure should
+be treated as code; they should be placed under source control and modularized
+in reusable components with automated testing and deployment.
+
+
+## Motivation
+
+The Tekton pipeline definitions are currently dissociated from a source code repository. 
+They may be located in the same repository as they code they should build, test and deploy,
+but Tekton is not aware of them. Tekton expects the pipeline definition to be deployed to a
+kubernetes cluster. Even when the Tekton resources are deployed to the cluster from the
+Repository, it's not easy to test changes to them, as their definition won't be updated until
+they are redeployed to the cluster. 
+
+The reason for the current behaviour is that Tekton is non-prescriptive about how to set up
+CI and CD pipelines. It aims to provide the building blocks for creating such pipelines and it leaves
+it up to the user to decide how to integrate the pipeline in the overall workflow. 
+
+The use case of a single repo, running CI/CD jobs in isolation, is very common and
+we wish to provide dedicated support for it, by providing an easy way for users to
```

Comment:
i think even stronger than that, although tekton doesn't want to be prescriptive, i think we CAN be prescriptive about strongly recommending (and easily supporting) ppl storing tasks + pipelines into version control (some more context and discussion in https://github.com/tektoncd/pipeline/issues/2298#issuecomment-725698929)

Reply:
I agree we can (and should), but I also think it really shouldn't be at the `tektoncd/pipeline` level 👼🏼

Reply:
fwiw, in our org we're using kubernetes namespaces to isolate pipelines and tasks which devs check into version control and that works fine for project level resources. We had to write additional components to achieve an e2e workflow though.

Reply:
I agree, we should have tasks and pipeline stored into version control, user can have some help with bundled and hopefully when we can have "remote task" if they need shared tasks.